pkg/workflow/controller/handlers/workflowTrigger/cronJob.go

```diff
@@ -0,0 +1,157 @@
+package workflowTrigger
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/caicloud/cyclone/pkg/k8s/clientset"
+	"github.com/pkg/errors"
+	"github.com/robfig/cron"
+	"github.com/satori/go.uuid"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/caicloud/cyclone/pkg/apis/cyclone/v1alpha1"
+)
+
+const (
+	KeyTemplate = "%s/%s"
+)
+
+type CronTrigger struct {
+	Cron                *cron.Cron
+	IsRunning           bool
+	SuccCount           int
+	FailCount           int
+	Namespace           string
+	WorkflowTriggerName string
+	WorkflowRun         *v1alpha1.WorkflowRun
+	Manage              *CronTriggerManager
+}
+
+type CronTriggerManager struct {
+	Client         clientset.Interface
+	CronTriggerMap map[string]*CronTrigger
+	mutex          sync.RWMutex
+}
+
+func NewTriggerManager(client clientset.Interface) *CronTriggerManager {
+	return &CronTriggerManager{
+		Client:         client,
+		CronTriggerMap: make(map[string]*CronTrigger),
+		mutex:          sync.RWMutex{},
+	}
+}
+
+func getKeyFromTrigger(trigger *CronTrigger) (string) {
+	return fmt.Sprintf(KeyTemplate, trigger.Namespace, trigger.WorkflowTriggerName)
+}
+
+func (m *CronTriggerManager) AddTrigger(trigger *CronTrigger) {
+	wftKey := getKeyFromTrigger(trigger)
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+	if wft, ok := m.CronTriggerMap[wftKey]; ok {
+		// this situation may not happen for k8s will do same name check
+		log.Warnf("failed to add cronTrigger, already exists: %+v\n", wft)
+	} else {
+		m.mutex.Lock()
+		m.CronTriggerMap[wftKey] = trigger
+		m.mutex.Unlock()
+	}
+}
+
+func (m *CronTriggerManager) DeleteTrigger(wftKey string) {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
```

Comment:
You are to `delete`, so it's not read lock.

Reply:
this is outdated